include/seastar/coroutine/try_future.hh

```diff
@@ -0,0 +1,135 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2025-present ScyllaDB
+ */
+
+#pragma once
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/util/noncopyable_function.hh>
+
+namespace seastar {
+
+namespace internal {
+
+using try_future_exception_handler_fn = noncopyable_function<std::exception_ptr(std::exception_ptr)>;
+
+template <bool CheckPreempt, typename T>
+class [[nodiscard]] try_future_awaiter {
+    seastar::future<T> _future;
+    try_future_exception_handler_fn _ex_handler;
+
+public:
+    explicit try_future_awaiter(seastar::future<T>&& f, try_future_exception_handler_fn ex_handler) noexcept : _future(std::move(f)), _ex_handler(std::move(ex_handler)) {}
+
+    try_future_awaiter(const try_future_awaiter&) = delete;
+    try_future_awaiter(try_future_awaiter&&) = delete;
+
+    bool await_ready() const noexcept {
+        return _future.available() && (!CheckPreempt || !need_preempt());
+    }
+
+    template<typename U>
+    void await_suspend(std::coroutine_handle<U> hndl) noexcept {
+        if (!CheckPreempt || !_future.available()) {
+            _future.set_coroutine(hndl.promise());
+        } else {
+            schedule(&hndl.promise());
+        }
+        hndl.promise().set_get_exception([this] () -> std::exception_ptr {
+            if (!_future.failed()) {
+                return {};
+            }
+            try {
+                if (_ex_handler) {
+                    return _ex_handler(_future.get_exception());
+                }
+                return _future.get_exception();
+            } catch (...) {
+                return std::current_exception();
+            }
+        });
```

Comment:
Perhaps a way to avoid the footprint increase of `promise` is to make the awaiter itself a seastar::task, schedule it self instead of the original coroutine (or call _future.set_coroutine on itself), inspect the received future, and decide whether to resume the coroutine or destroy it based on the result. Not at all sure it can work, this is really difficult.